### PR TITLE
Test tidy should not count untracked paths towards entries limit

### LINKED ISSUE #103951: `./x.py test tidy` should not count untracked paths towards entries limit
When browsing folders in macOS' Finder app, the OS automatically creates a `.DS_Store` file in the folder you're browsing. `.DS_Store` is usually included in `.gitignore` by users, making it trickier to detect which file increased the entries count in a given folder when getting the following error:

`tidy error: following path contains more than 948 entries, you should move the test to some relevant subdirectory (current: 949): …/rust/src/test/ui`

Because of scenarios like this, and because of the original purpose of the entries check described in https://github.com/rust-lang/rust/issues/73494, **`tidy` should ignore untracked files and directories inside tests directories.**

Related links:
- https://github.com/rust-lang/rust/pull/79959
- [Zulip topic - Tidy error: path contains more than 948 entries](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Tidy.20error.3A.20path.20contains.20more.20than.20948.20entries)
- https://github.com/rust-lang/rust/issues/75582

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #103951
r? @jyn514

### CODE REVIEW

**Comment by Nilstrieb** [comment]
What's the perf impact of this? Could you run `time x test tidy` before and after on your machine?

**Comment by chenyukang** [comment]
@Nilstrieb 

Old:
```
cat@oops:~/rust$ hyperfine --runs 5 'x test tidy'
Benchmark 1: x test tidy
  Time (mean ± σ):      8.154 s ±  1.040 s    [User: 24.769 s, System: 3.725 s]
  Range (min … max):    7.140 s …  9.679 s    5 runs
```

New:
```
cat@oops:~/rust$ hyperfine --runs 5 'x test tidy'
Benchmark 1: x test tidy
  Time (mean ± σ):      8.185 s ±  1.048 s    [User: 26.050 s, System: 4.057 s]
  Range (min … max):    7.370 s …  9.952 s    5 runs
```

**Comment by Nilstrieb** [comment]
Thanks, that looks great!

File: src/tools/tidy/src/ui_tests.rs [hunk 1]
```diff
@@ -11,35 +13,43 @@ const ROOT_ENTRY_LIMIT: usize = 948;
...
 fn check_entries(path: &Path, bad: &mut bool) {
-    let dirs = walkdir::WalkDir::new(&path.join("test/ui"))
-        .into_iter()
-        .filter_entry(|e| e.file_type().is_dir());
-    for dir in dirs {
-        if let Ok(dir) = dir {
-            let dir_path = dir.path();
+    for dir in Walk::new(&path.join("test/ui")) {
+        match dir {
+            Ok(entry) => {
```
Comment (jyn514) [inline src/tools/tidy/src/ui_tests.rs:18]: nit: changing this from `if let` to match makes the diff harder to understand

**Review by jyn514 — APPROVED** [review]
Thanks! I'm surprised how simple this is, nice work :) have you tested locally that it ignores untracked files in src/test/ui?

**Comment by chenyukang** [comment]
> Thanks! I'm surprised how simple this is, nice work :) have you tested locally that it ignores untracked files in src/test/ui?

Yes, I tested it locally.
I'm curious how this crate find out the `.gitignore`, it will find all the parents of a file/dir, 
https://github.com/BurntSushi/ripgrep/blob/000015791742bb1280f1853adb714fdee1ba9f8e/crates/ignore/src/dir.rs#L195

```console
add path: "/home/cat/rust/.gitignore"
dir_for_ignorefile: "/home/cat/rust/.git"
add path: "/home/cat/rust/.git/info/exclude"
dir_for_ignorefile: "/home/cat/rust/src"
dir_for_ignorefile: "/home/cat/rust/src"
dir_for_ignorefile: "/home/cat/rust/src/.git"
dir_for_ignorefile: "/home/cat/rust/src/test"
dir_for_ignorefile: "/home/cat/rust/src/test"
dir_for_ignorefile: "/home/cat/rust/src/test/.git"
dir_for_ignorefile: "/home/cat/rust/src/test/ui"
dir_for_ignorefile: "/home/cat/rust/src/test/ui"
```

But the performance seems OK.

File: src/tools/tidy/src/ui_tests.rs (hunk 1 above)
Comment (chenyukang) [reply src/tools/tidy/src/ui_tests.rs]: Add `if let`, but seems we can not use `let_chains`?

```console
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information
   = help: add `#![feature(let_chains)]` to the crate attributes to enable
```

And I and it to `lib.rs` also report:
```console
error[E0725]: the feature `let_chains` is not in the list of allowed features
 --> src/tools/tidy/src/lib.rs:1:12
  |
1 | #![feature(let_chains)]
  |            ^^^^^^^^^^
```

Comment (jyn514) [reply src/tools/tidy/src/ui_tests.rs]: ah, I see. this is fine then

**Pull request merged**